service/spawn.go

```diff
@@ -299,11 +299,37 @@ func (uis *UIServer) modifySpawnHost(w http.ResponseWriter, r *http.Request) {
 		return
 
 	case HostExpirationExtension:
+		if updateParams.Expiration.IsZero() { // set expiration to never expire
+			count, err := host.CountSpawnhostsWithNoExpirationByUser(u.Id)
+			if err != nil {
+				PushFlash(uis.CookieStore, r, w, NewErrorFlash("Error retrieving user spawn host data"))
+				uis.LoggedError(w, r, http.StatusInternalServerError, errors.Wrap(err, "Error retrieving user spawn host data"))
+				return
+			}
+			if count >= host.MaxSpawnhostsWithNoExpirationPerUser {
+				msg := fmt.Sprintf("Can only have %d non-expirable spawn hosts", host.MaxSpawnhostsWithNoExpirationPerUser)
+				PushFlash(uis.CookieStore, r, w, NewErrorFlash(msg))
+				uis.LoggedError(w, r, http.StatusBadRequest, errors.New(msg))
+				return
+			}
+			noExpiration := true
+			if err = cloud.ModifySpawnHost(ctx, h, host.HostModifyOptions{NoExpiration: &noExpiration}, env.Settings()); err != nil {
+				PushFlash(uis.CookieStore, r, w, NewErrorFlash("Error updating host expiration"))
+				uis.LoggedError(w, r, http.StatusInternalServerError, errors.Wrap(err, "Error extending host expiration"))
+				return
+			}
+			return
+		}
+		// use now as a base for how far we're extending if there is currently no expiration
+		if h.NoExpiration {
+			h.ExpirationTime = time.Now()
```

Comment:
My only concern with this new approach is that the time.Now() in the UI will be slightly different from the time.Now() here, so if people are trying to basically set it to expire immediately it'll say expiration can only be extended. I think that's maybe fine though.

Reply:
Could you use time 0 like before?

Reply:
I'm not sure I understand this comment. time.Now() is needed here to assert that the user hasn't chosen a time in the past / to calculate "additional hours" for the rest of the logic.